### PR TITLE
[data_collection_server] Add rosbag to data collection server

### PR DESCRIPTION
**Pull request by Kanazawanaoaki**
Rosbag can be used from data_collection_server corresponding to https://github.com/jsk-ros-pkg/jsk_common/issues/1707.  

You can also record rosbag when calling the service, by setting rosparam as shown below.
```
method: timer
rosbag_topics:
         - /rosout
         - /tf
         - /joint_states
```

## sample usage
1. Launch [this launch file](https://gist.github.com/Kanazawanaoaki/04810458e64c0a257a1f092379c17f0b#file-rosbag_in_data_collection_server_test-launch), when connecting to PR2.
```
wget https://gist.githubusercontent.com/Kanazawanaoaki/04810458e64c0a257a1f092379c17f0b/raw/6bd64995a723780424a938448d865f3d33bf690d/rosbag_in_data_collection_server_test.launch
roslaunch rosbag_in_data_collection_server_test.launch
```
2. Call `start_request` to start data collection and record rosbag.
```
rosservice call /data_collection_rosbag_test/start_request "{}"
```
3. Call `end_request` to stop data collection and stop rosbag recording.
```
rosservice call /data_collection_rosbag_test/end_request "{}"
```
rosbag file will be saved in `~save_dir`
```
cd ~/Desktop/data_collection_rosbag_test/
ls
```

### CODE REVIEW

**Comment by shmpwk** [comment]
Thank you for your work. LGTM.

File: jsk_data/node_scripts/data_collection_server.py [hunk 1]
```diff
@@ -187,6 +193,19 @@ def sub_cb(self, msg, topic_name):
             'msg': msg
         }
 
+    def start_rosbag(self):
+        postfix = rospy.Time.now()
+        filename = os.path.join(
+            self.save_dir, 'rosbag-{0}'.format(postfix))
```
Comment (knorth55) [inline jsk_data/node_scripts/data_collection_server.py:199]: how about adding rosbag prefix option to param?
```
rosbag_prefix: baxter_rosbag_
```

**Review by knorth55 — APPROVED** [review]

File: jsk_data/node_scripts/data_collection_server.py [hunk 2]
```diff
@@ -91,6 +93,10 @@ def __init__(self):
         self.timestamp_save_dir = rospy.get_param('~timestamp_save_dir', True)
         self.wait_timer = rospy.get_param('~wait_timer', False)
         self.wait_save_request = rospy.get_param('~wait_save_request', False)
+        self.rosbag = rospy.get_param('~rosbag', False)
+        if self.rosbag:
+            self.rosbag_process = None
+            self.rosbag_topics = rospy.get_param('~rosbag_topics', [])
```
Comment (shmpwk) [inline jsk_data/node_scripts/data_collection_server.py:99]: simpler?
```
        self.rosbag_topics = rospy.get_param('~rosbag_topics', [])
        if self.rosbag_topics:
            self.rosbag_process = None
```

File: jsk_data/node_scripts/data_collection_server.py [hunk 3]
```diff
@@ -277,10 +296,14 @@ def _save(self):
 
     def start_service_cb(self, req):
         self.start = True
+        if self.rosbag:
```
Comment (shmpwk) [inline jsk_data/node_scripts/data_collection_server.py:299]: simpler?
```
        if self.rosbag_topics:
```

File: jsk_data/node_scripts/data_collection_server.py [hunk 4]
```diff
@@ -277,10 +296,14 @@ def _save(self):
...
+            self.start_rosbag()
         return TriggerResponse(success=True)
 
     def end_service_cb(self, req):
         self.start = False
+        if self.rosbag:
```
Comment (shmpwk) [inline jsk_data/node_scripts/data_collection_server.py:305]: simpler?
```
        if self.rosbag_topics:
```

File: jsk_data/node_scripts/data_collection_server.py (hunk 1 above)
Comment (knorth55) [reply jsk_data/node_scripts/data_collection_server.py:199]:
```
<launch>
  <arg name="collect_data" default="true" />

  <node name="data_collection_rosbag_test"
        pkg="jsk_data" type="data_collection_server.py"
        output="screen" if="$(arg collect_data)">
    <rosparam subst_value="true">
      message_filters: true
      approximate_sync: true
      save_dir: ~/Desktop/data_collection_rosbag_test
      method: timer
      topics:
          - name: /kinect_head/rgb/image_rect_color
            msg_class: sensor_msgs/Image
            fname: pr2_image_rect_color_request.png
            savetype: ColorImage
      rosbag: true
      rosbag_prefix: pr2_rosbag_
      rosbag_topics:
           - /rosout
           - /tf
           - /tf_static
           - /joint_states
           - /map
           - /base_odometry/odom
           - /robot_pose_ekf/odom_combined
           - /base_controller/command
           - /navigation/cmd_vel
           - /move_base_node/NavFnROS/plan
           - /move_base_node/DWAPlannerROS/global_plan
           - /move_base_node/DWAPlannerROS/local_plan
           - /move_base_node/local_costmap/costmap
           - /move_base_node/global_costmap/costmap
           - /move_base_node/global_costmap/footprint
           - /safe_teleop_base/local_costmap/costmap
           - /spots_marker_array
           - /particlecloud
           - /base_scan_throttled
           - /tilt_scan_throttled
           - /kinect_head/rgb/throttled/camera_info
           - /kinect_head/depth_registered/throttled/camera_info
           - /kinect_head/rgb/throttled/image_rect_color/compressed
           - /kinect_head/depth_registered/throttled/image_rect/compressedDepth
           - /audio
    </rosparam>
  </node>  
</launch>
```

**Review by shmpwk — APPROVED** [review]

File: jsk_data/node_scripts/data_collection_server.py (hunk 2 above)
Comment (iory) [inline jsk_data/node_scripts/data_collection_server.py:99]: Hou about adding ```rospy.get_param('~rosbag_options', [])```?

**Review by iory — CHANGES_REQUESTED** [review]

File: jsk_data/node_scripts/data_collection_server.py (hunk 1 above)
Comment (Kanazawanaoaki) [reply jsk_data/node_scripts/data_collection_server.py:199]: Thank you. nice one.
I would like to add rosbag prefix option to param.

File: jsk_data/node_scripts/data_collection_server.py (hunk 4 above)
Comment (Kanazawanaoaki) [reply jsk_data/node_scripts/data_collection_server.py:305]: Thank you. is that so.
I will remove the param of rosbag and change it to take rosbag if rosbag_topics is specified.

File: jsk_data/node_scripts/data_collection_server.py (hunk 2 above)
Comment (Kanazawanaoaki) [reply jsk_data/node_scripts/data_collection_server.py:99]: Thank you for your review.
I don't understand it well, but is it to integrate the rosbag type params (like `rosbag_topics` and `rosbag_prefix`) into `rosbag_options` param?

**Comment by Kanazawanaoaki** [comment]
`rosbag_prefix` param was added, and the redundant param of `rosbag` was abolished.  
I also updated gist and https://github.com/jsk-ros-pkg/jsk_common/pull/1708#issue-1096247724.

File: jsk_data/node_scripts/data_collection_server.py (hunk 2 above)
Comment (iory) [reply jsk_data/node_scripts/data_collection_server.py]: We can pass an option to the rosbag record, so it would be nice if the user could pass an option.

Comment (Kanazawanaoaki) [reply jsk_data/node_scripts/data_collection_server.py]: I see.
I'm wondering if options related to topics and file names such as -a and -o should be supported, but I made it possible to specify the option of rosbag record from `rosbag_options` param.

Comment (iory) [reply jsk_data/node_scripts/data_collection_server.py]: `rosbag record` has a lot of options, and other jsk's programs use such that options. So, it's good features. Thanks.
http://wiki.ros.org/rosbag/Commandline#record
https://github.com/jsk-ros-pkg/jsk_common/blob/master/jsk_data/launch/rosbag_always_run_rosbag.launch#L7

**Review by iory — APPROVED** [review]

**Pull request merged**